crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -0,0 +1,274 @@
+use std::cmp::Ordering;
+
+use crate::types::Type;
+
+use super::{
+    class_base::ClassBase, ClassLiteralType, DynamicType, InstanceType, KnownInstanceType, TodoType,
+};
+
+/// Return an [`Ordering`] that describes the canonical order in which two types should appear
+/// in an [`crate::types::IntersectionType`] or a [`crate::types::UnionType`] in order for them
+/// to be compared for equivalence.
+///
+/// Two unions with equal sets of elements will only compare equal if they have their element sets
+/// ordered the same way.
+///
+/// ## Why not just implement `Ord` on `Type`?
+///
+/// It would be fairly easy to slap `#[derive(PartialOrd, Ord)]` on `Type`. However, this would
+/// order types according to their Salsa ID. While this would mean that types would always be
+/// consistently ordered in any single run of red-knot, the order in which they would appear
+/// might vary between different runs of red-knot. The ordering we create here is not user-facing;
+/// however, it will be useful in the future to know where, for example, any `Protocol` elements
+/// lie in the union's element list (if there are any `Protocol` elements in the union's element
+/// list).
+///
+/// Moreover, it doesn't really "make sense" for `Type` to implement `Ord` in terms of the
+/// semantics. There are many different ways in which you could plausibly sort a list of types;
+/// this is only one (somewhat arbitrary, at times) possible ordering.
+pub(super) fn order_union_elements<'db>(left: &Type<'db>, right: &Type<'db>) -> Ordering {
```

Comment:
Same as I'd not call this `sort_` I'd also not call this `order` because it doesn't reorder elements. I'd instead just call it `xy_ordering`

```suggestion
pub(super) fn union_elements_ordering<'db>(left: &Type<'db>, right: &Type<'db>) -> Ordering {
```